### ISSUE TITLE
Horovod applied to a LSTM

### ISSUE THREAD

**Issue by JohnTaylor2000** [horovod/horovod#1833]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet) any
2. Framework version: any
3. Horovod version: any
4. MPI version: any
5. CUDA version: any
6. NCCL version: any
7. Python version: any
8. OS and version: any
9. GCC version: any

**Checklist:**
1. Did you search issues to find if somebody asked this question before? Yes
2. If your question is about hang, did you read [this doc] (https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)? Yes

**Your question:**
Please ask your question here.

Does anyone have an example of how to apply horovod to an LSTM (single time series)  problem ideally using keras/tensorflow?

**Comment by abditag2** [comment]
There is not any difference for using LSTM vs. other types of models. You can follow any of our examples and only swap out the model with your model and inject your data.

**Comment by JohnTaylor2000** [comment]
Yes you are correct there are many ways to achieve this based on exisiting applications of horovod.  Time series problems do pose their own unique problems. I am interested identifying the most efficient way of using horovod ie how the LSTM problems map to solutions using horovod, so was interested in surveying the approaches that others have taken, but have not found many LSTM examples using horovod. So if anyone has an example application I would interested to learn  more about that application.

**Comment by tgaddair** [comment]
Hey @JohnTaylor2000, [OpenSeq2Seq](https://devblogs.nvidia.com/mixed-precision-nlp-speech-openseq2seq/) is one example that combines LSTM (among other layers/cells) with Horovod.  However, the way they use Horovod is pretty raw, and I wouldn't recommend it in most cases.

[Ludwig](https://github.com/uber/ludwig) is another framework that uses Horovod and is commonly applied to NLP tasks.

For most people, I would just recommend following the pattern of using `hvd.DistributedOptimizer` shown in the [examples](https://github.com/horovod/horovod/tree/master/examples).  Under the hood, this class will allreduce your gradients each step (or allgather in the case of Indexed Slices).

Let me know if you have any specific concerns we can help address.  The beauty of Horovod is that it doesn't require much work from the ML engineer to be efficient.  The only efficiency concerns come in when we start talking about hardware and other infrastructure components.

**Comment by kingwales1** [comment]
Hi @JohnTaylor2000 Did you figure it out?

Also, @tgaddair like the LSTM training example from [tensorflow](https://github.com/tensorflow/models/blob/master/research/lstm_object_detection/train.py), can I follow the pattern of using hvd.DistributedOptimizer to distribute it?

**Comment by JohnTaylor2000** [comment]
Hi kingwales1,

You can apply horovod as per the documentation by using using hvd.DistributedOptimizer.

I have gone back to a single GPU implementation to sort out efficiency issues. These issues are definitely problem specific so no single recipe here. For my problem I have found that generator methods are useful. The quality and time to solution were most sensitive to batch size (1024-2048), sequence length (4-32) and number of hidden units (2x sequence length), but keep in mind that your problem may respond very differently.

Best wishes,

John
